### DISCUSSION TITLE
post/ebpf-talk-29-failed-to-xdp-redirect/

### DISCUSSION THREAD

**Discussion by giscus** [Asphaltt/asphaltt.github.io#1] [Announcements]
# post/ebpf-talk-29-failed-to-xdp-redirect/

奇葩场景遇到个奇葩问题。 为了更好的性能，就将 XDP 程序挂载到网卡驱动里。但有个业务需求，在 XDP 程序里将需要延迟的流量转发到 veth 设备。所以，就直接在 XDP

https://asphaltt.github.io/post/ebpf-talk-29-failed-to-xdp-redirect/

**Comment by vincent5753** [comment]
您好，想請教一下
我最近想使用基於 XDP 實作簡易的 LB (環境是位於 Kubernetes 上的 veth)
我想達成 A -> C 時改寫並轉送至 B (在 veth 上)
對於 B 而言封包會像是 A -> B
而我在更改 IP 和 MAC 位置後
使用 bpf_redirect、XDP_TX 或 XDP_REDIRECT
B 都沒辦法收到封包
使用 perf 或是 bpftrace 會得到
# bpf_redirect(ifindex2, 0)
xdp:xdp_redirect_err:prog_id=1426 action=REDIRECT ifindex=66 to_ifindex=67 err=-6
# XDP_TX
xdp:xdp_bulk_tx:ifindex=66 action=TX sent=0 drops=1 err=-6
# XDP_REDIRECT
xdp:xdp_redirect_err:prog_id=1428 action=REDIRECT ifindex=66 to_ifindex=0 err=-22
有嘗試過把 NAPI 開啟並把 tx checksum 關閉，但還是沒有成功
想情問能給一些除錯方向或是指教嗎？

**Reply by Asphaltt** [reply]
简单来说，需要 `ethtool -K veth-B gro on`；此时，veth-B 才能接收 XDP redirect 过来的网络包。

**Reply by vincent5753** [reply]
您好，我嘗試過在 host 執行過 `ethtool -K vethxxx gro on` 並且也在容器內執行過 `ethtool -K eth0 gro on`
也嘗試在容器的網路介面(veth)上掛載簡單的 XDP_PASS 的 dummp XDP
參考: [Re: Veth pair swallow packets for XDP_TX operation](https://www.spinics.net/lists/netdev/msg625217.html)
但我的 XDP_TX 轉送一樣是沒有成功
想請問您能提供一些指引或是和告訴我您的開發環境讓我去試嗎？

我的環境如下
OS: Ubuntu 20.04
Kernel: 5.4.0-144-generic
Docker: 20.10.9

**Reply by Asphaltt** [reply]
能否提供具体的网络拓扑？

比如：
1. veth-A (in Docker) --veth-A (on host)-> XDP redirect --veth-B (on host)-> veth-B (in Docker)
2. eth-A (NIC on host) --> XDP redirect --veth-B (on host)-> veth-B (in Docker)

**Reply by vincent5753** [reply]
我的環境是直接使用 Docker run 建立容器所產生的 veth(於 host 上)
所以如果我的理解沒錯應該是如下
Aeth0(容器內) -> Aveth(host) -> XDP redirect / TX -> Bveth(host) -> Beth0(容器內)

我建立環境的步驟如下
docker run --privileged --name=nginx1 -d nginx:latest
docker run --privileged --name=nginx2 -d nginx:latest
docker run --privileged --name=lb -d nginx:latest
docker run --privileged --name=client -d nginx:latest
# 在容器建立後(分別對四個容器操作)
docker exec ${容器名稱} ethtool -K eth0 gro on
sudo ethtool -K ${vethxxxx} gro on

# 並掛載 XDP
# 掛載至 LB
sudo ip link set dev vetha2b7ca7 xdpdrv obj xdp_lb_kern.o sec xdp_lb
# 掛載至其他容器(內容只有 XDP_PASS)
sudo ip link set dev veth577651b xdpdrv obj justpass.o sec xdpprogram

其他容器需要這樣掛載的原因我是參照這篇
https://www.spinics.net/lists/netdev/msg625217.html#:~:text=You%20need%20to%20load%20a,when%20using%20native%20veth%20XDP.

**Reply by Asphaltt** [reply]
`docker run --privileged --name=lb -d nginx:latest` 没有使用 host network？

client 的流量是怎么到达 lb 的？lb 的流量又是怎么负载到 nginx1, nginx2 的？你自己弄明白这个问题后，事情就简单很多了。

TIP: 或许可以使用 [pwru](github.com/cilium/pwru) 观察一下流量从哪里到了哪里。
